// @nxrocks/nx-spring-boot/src/generators/project/lib/generate-project-configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateProjectConfiguration = void 0;
const devkit_1 = require("@nx/devkit");
const common_jvm_1 = require("@nxrocks/common-jvm");
const plugin_utils_1 = require("../../../utils/plugin-utils");
function generateProjectConfiguration(tree, options) {
    devkit_1.logger.info(`⚙️ Generating project configuration...`);
    if (options.transformIntoMultiModule) {
        (0, devkit_1.addProjectConfiguration)(tree, options.parentModuleName, {
            root: options.moduleRoot,
            sourceRoot: `${options.moduleRoot}`,
            projectType: options.projectType,
            targets: {},
            tags: options.parsedTags,
        });
    }
    (0, devkit_1.addProjectConfiguration)(tree, options.projectName, Object.assign(Object.assign({ root: options.projectRoot, sourceRoot: (0, devkit_1.joinPathFragments)(options.projectRoot, 'src') }, (!(0, common_jvm_1.isNxCrystalEnabled)()
        ? (0, plugin_utils_1.getProjectTypeAndTargetsFromOptions)(options)
        : {})), { tags: options.parsedTags }));
}
exports.generateProjectConfiguration = generateProjectConfiguration;
