// @nxrocks/nx-spring-boot/src/generators/project/lib/normalize-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeOptions = void 0;
const tslib_1 = require("tslib");
const project_name_and_root_utils_1 = require("@nx/devkit/src/generators/project-name-and-root-utils");
function normalizeOptions(tree, options) {
    var _a;
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { projectName, projectRoot, projectNameAndRootFormat, } = yield (0, project_name_and_root_utils_1.determineProjectNameAndRootOptions)(tree, {
            name: options.name,
            projectType: options.projectType,
            directory: options.directory,
            projectNameAndRootFormat: options.projectNameAndRootFormat,
            //rootProject: options.rootProject,
            callingGenerator: '@nxrocks/nx-spring-boot:project',
        });
        options.projectNameAndRootFormat = projectNameAndRootFormat;
        const parsedTags = options.tags
            ? options.tags.split(',').map((s) => s.trim())
            : [];
        const projectDependencies = ((_a = options.dependencies) === null || _a === void 0 ? void 0 : _a.split(',').map((s) => s.trim()).filter((s) => !!s)) || [];
        return Object.assign(Object.assign({}, options), { projectName,
            projectRoot,
            projectDependencies,
            parsedTags });
    });
}
exports.normalizeOptions = normalizeOptions;
