// @nxrocks/nx-spring-boot/src/generators/project/lib/prompt-boot-dependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.promptBootDependencies = void 0;
const tslib_1 = require("tslib");
const enquirer_1 = require("enquirer");
const boot_utils_1 = require("../../../utils/boot-utils");
const devkit_1 = require("@nx/devkit");
function promptBootDependencies(options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (options.dependencies === undefined &&
            process.env.NX_INTERACTIVE === 'true') {
            devkit_1.logger.info(`â³ Fetching Spring Boot dependencies list. Please wait...`);
            const dependencies = Object.keys(yield (0, boot_utils_1.fetchBootDependencies)(options));
            options.projectDependencies = yield (0, enquirer_1.prompt)({
                name: 'dependencies',
                message: 'What dependencies would you like to use? (type something to filter, press [space] to multi-select)',
                type: 'autocomplete',
                choices: dependencies,
                multiple: true,
            }).then((a) => a['dependencies']);
        }
    });
}
exports.promptBootDependencies = promptBootDependencies;
